packages/internal/src/types.d.ts

```diff
@@ -18,3 +18,10 @@ export declare class SyncCallback<
 
   public isSync: true;
 }
+
+declare global {
+  /**
+   * Utility for the type of an Exo object once prepared and instantiated.
+   */
+  type ExoObj<T> = Awaited<ReturnType<Awaited<ReturnType<T>>>>;
```

Comment:
With a little recursion, this becomes useful not just for Exos:
```suggestion
  /**
   * Find the eventual return type of a specimen, recursing to function
   * return values and promise fulfilments.
   */
  type EReturn<T> = T extends (...args: any[]) => infer R ? EReturn<R> : 
    T extends PromiseLike<infer U> ? EReturn<U> : 
    T;
```

Reply:
That would be more general, yes. I think one for Exos has some didactic value so even if we had this I'd suggest having both.

What use cases do you see for a general `EReturn`?

Reply:
On second thought, since having `EReturn` would cover all the cases I don't think it's worth naming another type.

@kriskowal would you be ok with an `EReturn` type in Endo? You expressed some concern earlier about recursive types.

```ts
const makeHi = () => Promise.resolve('hi' as const);
type Hi = EReturn<typeof makeHi>;
const hi = await makeHi();
expectType<'hi'>(hi);
expectType<Hi>(hi);
```

Reply:
Is this also the case?
```ts
const lol = async () => async () => async () => 'lol';
type LoL = EReturn<typeof lol>;
expectType<'lol'>(LoL);
```
That would seem weird to me, recursively unwrapping both promise and return type. I would think it safer to be more specific than an arbitrary number of niladic applications.

Is it possible to have appropriate type extractors with names that resemble the corresponding constructors? I like having `Return` in the name. I recall @erights had a table with all the Exo constructor patterns named. Maybe that would help inform a family of return-type-of operator names.

Reply:
That would be the case, yes.

> I would think it safer to be more specific than an arbitrary number of niladic applications.

What hazard(s) are you defending against?

>Is it possible to have appropriate type extractors with names that resemble the corresponding constructors? 

Certainly. It's just more to write, maintain and require people to learn.

Reply:
I would expect `EReturn<T>` to be shallow, so it could be used to eventually capture the next function signature. It would be useful for remote functions. Can suggest reserving and pitch `EUltimateReturn<T>`?

Reply:
https://github.com/endojs/endo/pull/2659